### ISSUE TITLE
YOLO Hyperparameter Tuning and Data Augmentation

### ISSUE THREAD

**Issue by 123wula** [ultralytics/ultralytics#21627]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

My questions:

1. I am training a project based on yolov10n.pt. According to some practical suggestions, I first used the `default.yaml` settings and the pre-trained weights `yolov10n.pt` for training on a general basis, and now the training is halfway through.
2. Based on practical suggestions, I should try hyperparameter tuning and data augmentation to improve my model. However, I'm not sure whether I should directly overwrite the old parameters with the new hyperparameters after interruption and then resume training.I have found some potential points needing optimization.
3. I have consulted relevant materials but failed to obtain definite practical guidance and case examples.
If I can't interrupt the process for data augmentation or hyperparameter tuning, it may mean that I have to conduct numerous retraining sessions because I have to abandon the model that needs improvement.
4. I'm not clear about which models should be abandoned rather than tuned. For example, in the case of overfitting, is it necessary to abandon the model? Or, for some relatively low performance metrics (when training is already completed), should the model be abandoned?
5. I apologize for asking such basic questions, but I really don't know what is permissible in practical applications. I hope I can get some guidance! 😊

**Comment by UltralyticsAssistant** [comment]
👋 Hello @123wula, thanks for opening an issue and for your interest in Ultralytics 🚀 This is an automated response to help you get started quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the Docs for practical guidance and examples: check out the Python and CLI usage pages for end-to-end training, resuming, and tuning workflows:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please share as much detail as you can:
- Your training command or Python script
- hyperparameters used (including any changes from defaults)
- augmentation settings
- a few representative dataset images
- training logs/plots
And double-check you’re following our Tips for Best Training Results:
- Training Tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat 🎧: https://discord.com/invite/ultralytics
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit for threads and sharing: https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package and verify all requirements in a Python>=3.8 environment with PyTorch>=1.8 to ensure you’re on a current version before proceeding:

```bash
pip install -U ultralytics
```

Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To move your discussion forward, please clarify:
1) Whether this is a bug or a training/usage question
2) Your exact training/resume commands and hyperparameters
3) What augmentations you’re considering
4) Current metrics/curves and any signs of over/underfitting

Once we have these details, we can provide targeted guidance on hyperparameter tuning, augmentation strategies, and when to resume vs. restart training 😊

**Comment by Y-T-G** [comment]
You can't resume a training with different parameters. Resuming only continues with old parameters except batch size. You can start a new training with the old model's checkpoint that was saved. But this resets the optimizer, so it's not the same as resuming. It's just starting from those weights.

Hyperparameter tuning is perfomed on new training. It's a slow process and takes days to weeks to get something optimal.

Augmentations are automatically applied in Ultralytics. So you don't need to perform it manually.

**Comment by 123wula** [comment]
@Y-T-G Thank you for your reply! 😊   I've learned about the operations supported at checkpoint, which saves me the trouble of modifying the source code manually to update hyperparameter settings later.

Although with your help, I understand that these modifications are not supported, some data augmentation techniques only change the data reading process (the reading operation should be rerun every epoch) rather than the model parameters and optimizer parameters. In this case, it seems that data augmentation operations after interruption should be supported... 💡 Are there other reasons for this issue?

(Some anecdotes:
Previously, I had doubts about this but couldn't find any materials related to this problem. So, I had to try modifying the code since methods like passing parameters didn't work. I manually modified settings in get_cfg() in ultralytics\cfg_init_.py, such as cfg['mosaic']=0.0. I saw the modification in the printed information, but cfg['lr0']=0.0003 actually didn't change the learning rate. Maybe iterative updates inside the _do_train function are required...)Perhaps its design is intended for safety reasons...？
